src/vulkan_as_3.cpp

```diff
@@ -0,0 +1,339 @@
+			vertices.size() * sizeof(Vertex), 
+			vertices.data(), 
+			VK_BUFFER_USAGE_SHADER_DEVICE_ADDRESS_BIT | VK_BUFFER_USAGE_ACCELERATION_STRUCTURE_BUILD_INPUT_READ_ONLY_BIT_KHR, 
+			VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT | VK_MEMORY_PROPERTY_HOST_COHERENT_BIT
+		);
+		resources.vertex_buffers[as_index].address.deviceAddress = acceleration_structures::get_buffer_device_address(vulkan, resources.vertex_buffers[as_index].handle);
+
+		resources.index_buffers[as_index] =  acceleration_structures::prepare_buffer(
+			vulkan, 
+			indices.size() * sizeof(uint32_t), 
+			indices.data(), 
+			VK_BUFFER_USAGE_SHADER_DEVICE_ADDRESS_BIT | VK_BUFFER_USAGE_ACCELERATION_STRUCTURE_BUILD_INPUT_READ_ONLY_BIT_KHR, 
+			VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT | VK_MEMORY_PROPERTY_HOST_COHERENT_BIT
+		);
+		resources.index_buffers[as_index].address.deviceAddress = acceleration_structures::get_buffer_device_address(vulkan, resources.index_buffers[as_index].handle);
+	}
+}
+
+void free_test_resources(const vulkan_setup_t& vulkan, Resources & resources){
+    vkDestroyQueryPool(vulkan.device, resources.query_pool, nullptr);
+	vkDestroyCommandPool(vulkan.device, resources.command_pool, nullptr);
+	
+    for(uint32_t as_index = 0; as_index < as_build_count; ++as_index){
+        resources.functions.vkDestroyAccelerationStructure(vulkan.device, resources.bl_acc_structures[as_index].handle, nullptr);
+        vkFreeMemory(vulkan.device, resources.original_blas_buffers[as_index].memory, nullptr);
+        vkDestroyBuffer(vulkan.device, resources.original_blas_buffers[as_index].handle, nullptr);
+    
+        resources.functions.vkDestroyAccelerationStructure(vulkan.device, resources.opt_bl_acc_structures[as_index].handle, nullptr);
+	    vkFreeMemory(vulkan.device, resources.optimized_blas_buffers[as_index].memory, nullptr);
+	    vkDestroyBuffer(vulkan.device, resources.optimized_blas_buffers[as_index].handle, nullptr);
+	
+	    vkFreeMemory(vulkan.device, resources.vertex_buffers[as_index].memory, nullptr);
+	    vkDestroyBuffer(vulkan.device, resources.vertex_buffers[as_index].handle, nullptr);
+	
+	    vkFreeMemory(vulkan.device, resources.index_buffers[as_index].memory, nullptr);
+	    vkDestroyBuffer(vulkan.device, resources.index_buffers[as_index].handle, nullptr);
+    }
+}
+
+void optimize_acceleration_structure(const vulkan_setup_t& vulkan, Resources & resources)
+{
+	const uint32_t num_triangles = 1;
+	std::vector<Buffer> as_scratch_buffers(as_build_count);
+	std::vector<VkAccelerationStructureGeometryKHR> as_geometries(as_build_count);
+	std::vector<VkAccelerationStructureBuildSizesInfoKHR> as_build_size_infos(as_build_count);
+	std::vector<VkAccelerationStructureBuildGeometryInfoKHR> as_build_geometry_infos(as_build_count);
+	std::vector<VkAccelerationStructureBuildRangeInfoKHR*> as_build_range_infos(as_build_count);
+	std::vector<VkAccelerationStructureKHR> as_handles(as_build_count);
+
+	for(uint32_t as_index = 0; as_index < as_build_count; ++as_index)
+    {
```

Comment:
I see some whitespace issues like this, although it is hard to see clearly what is happening in github review. Could you check and adjust your editor to make sure trailing whitespace is shown (and removed)?